### ISSUE TITLE
[Feature] Add FlexAttention version STA support

### ISSUE THREAD

**Issue by shadow150519** [hao-ai-lab/FastVideo#224]
### Motivation

Hi, thanks for your awesome project, I find now you only support one kind of size STA. Could you opensource flexattention version of STA implementation for easy of use?

**Comment by wangshankun** [comment]
test case里面有例子，自己简单改造一下，应该就行。  
https://github.com/hao-ai-lab/FastVideo/blob/main/csrc/sliding_tile_attention/test/test_sta.py#L50
 pt_o = flex_test(Q, K, V, kernel_size)

@jzhang38 ，我个人小提议，比起来用h100_fwd_kernel提升的部分性能，而牺牲flex attention的通用性有点得不偿失； flex attention和h100_fwd_kernel没必要非得二选一，对于开源项目来说，通用性和易用性的优先级可能会更高点。

**Comment by tomyoung903** [comment]
The current flex version of STA seems to be _**much slower**_ than the tk-based version. 

Using a script based on (csrc/sliding_tile_attention/test/bench.py and flex_sta_ref.py) to compare them

on kernel_size = [1, 6, 6], tk-based is **twice** as fast.
on kernel_size = [1, 1, 6], tk-based is **6 times** as fast.

Is that to be expected @BrianChen1129 @wangshankun ? . **The** paper says the tk-version is only 50% faster.

**Comment by shadow150519** [comment]
> The current flex version of STA seems to be _**much slower**_ than the tk-based version.
> 
> Using a script based on (csrc/sliding_tile_attention/test/bench.py and flex_sta_ref.py) to compare them
> 
> on kernel_size = [1, 6, 6], tk-based is **twice** as fast. on kernel_size = [1, 1, 6], tk-based is **6 times** as fast.
> 
> Is that to be expected [@BrianChen1129](https://github.com/BrianChen1129) [@wangshankun](https://github.com/wangshankun) ? . **The** paper says the tk-version is only 50% faster.

Hi,  how you configure tile_thw, block_thw in flexattention sta? I am use sta to replace flash attention on Wanx t2v 1.3b, for a latent vector of (21,30,52), I set tile_thw = (2,2,4) block_thw=(4,4,8). But it only generate a bunch of mosaics. 
<img width="757" alt="Image" src="https://github.com/user-attachments/assets/af6668a3-df86-4541-aa4e-2ec3eb52c6d1" />
In STA paper experiments, for a latent vector(30,48,80), sta uses two types config (18,24,24), (30,40,40). Could you tell me  which tile_size and kernel size do you use? @jzhang38 @BrianChen1129 @wangshankun 

<img width="799" alt="Image" src="https://github.com/user-attachments/assets/b0f4532e-079e-4912-858d-12dd0619bbda" />

following is my code implementation, `get_sliding_tile_attention_mask` is copied from [flex_sta_ref.py](https://github.com/hao-ai-lab/FastVideo/blob/main/csrc/sliding_tile_attention/test/flex_sta_ref.py)
```python
def sta_attention(
    q,
    k,
    v,
    grid_size,
    tile_size=(2,2,4),
    kernel_size=(4,4,8)
):
    q = q.to(v.dtype).transpose(1,2).contiguous()
    k = k.to(v.dtype).transpose(1,2).contiguous()
    v = v.transpose(1,2).contiguous()

    mask = get_sliding_tile_attention_mask(kernel_size, tile_size, grid_size, 'cuda')
    out = flex_attention(q, k, v,block_mask=mask)
    out = out.transpose(1, 2).contiguous()
    
    return out
```

**Comment by wangshankun** [comment]
@shadow150519 hunyuan的是改成下面这样，stepvideo可以简单替换；但是hunyuan有has_text，flex_attention没办法直接简单完成；因为text这个要单独处理逻辑；

def flex_test(Q, K, V, window_size, has_text):
    hidden_states = torch.empty_like(Q)
    for head_index, (t_kernel, h_kernel, w_kernel) in enumerate(window_size):
        q_head, k_head, v_head, o_head = Q[:, head_index:head_index +
                                            1], K[:, head_index:
                                                        head_index +
                                                        1], V[:,
                                                                head_index:
                                                                head_index +
                                                                1], hidden_states[:,
                                                                                    head_index:
                                                                                    head_index
                                                                                    +
                                                                                    1]
        mask = get_sliding_tile_attention_mask((**t_kernel, h_kernel, w_kernel), (6, 8, 10), (30, 48, 80)**, 39, 'cuda', 0)
        hidden_states[:,head_index:head_index + 1] = flex_attention(q_head, k_head, v_head, block_mask=mask)

    return hidden_states

**Comment by wangshankun** [comment]
> The current flex version of STA seems to be _**much slower**_ than the tk-based version.
> 
> Using a script based on (csrc/sliding_tile_attention/test/bench.py and flex_sta_ref.py) to compare them
> 
> on kernel_size = [1, 6, 6], tk-based is **twice** as fast. on kernel_size = [1, 1, 6], tk-based is **6 times** as fast.
> 
> Is that to be expected [@BrianChen1129](https://github.com/BrianChen1129) [@wangshankun](https://github.com/wangshankun) ? . **The** paper says the tk-version is only 50% faster.

@tomyoung903 
对的，不到50%，
<img width="643" alt="Image" src="https://github.com/user-attachments/assets/69ccb18a-170f-4c8c-951a-5d0d14bc53ea" />

**Comment by shadow150519** [comment]
> [@shadow150519](https://github.com/shadow150519) hunyuan的是改成下面这样，stepvideo可以简单替换；但是hunyuan有has_text，flex_attention没办法直接简单完成；因为text这个要单独处理逻辑；
> 
> def flex_test(Q, K, V, window_size, has_text): hidden_states = torch.empty_like(Q) for head_index, (t_kernel, h_kernel, w_kernel) in enumerate(window_size): q_head, k_head, v_head, o_head = Q[:, head_index:head_index + 1], K[:, head_index: head_index + 1], V[:, head_index: head_index + 1], hidden_states[:, head_index: head_index + 1] mask = get_sliding_tile_attention_mask((**t_kernel, h_kernel, w_kernel), (6, 8, 10), (30, 48, 80)**, 39, 'cuda', 0) hidden_states[:,head_index:head_index + 1] = flex_attention(q_head, k_head, v_head, block_mask=mask)
> 
> ```
> return hidden_states
> ```

In Wanx, I only apply sta on self-attention in which qkv are both video, so it's more like stepvideo rather than hunyuan. Have you ever tried other numbers for tile_thw, kernel_thw? I find #220 add some heuristic method to choose config.

**Comment by wangshankun** [comment]
> > [@shadow150519](https://github.com/shadow150519) hunyuan的是改成下面这样，stepvideo可以简单替换；但是hunyuan有has_text，flex_attention没办法直接简单完成；因为text这个要单独处理逻辑；
> > def flex_test(Q, K, V, window_size, has_text): hidden_states = torch.empty_like(Q) for head_index, (t_kernel, h_kernel, w_kernel) in enumerate(window_size): q_head, k_head, v_head, o_head = Q[:, head_index:head_index + 1], K[:, head_index: head_index + 1], V[:, head_index: head_index + 1], hidden_states[:, head_index: head_index + 1] mask = get_sliding_tile_attention_mask((**t_kernel, h_kernel, w_kernel), (6, 8, 10), (30, 48, 80)**, 39, 'cuda', 0) hidden_states[:,head_index:head_index + 1] = flex_attention(q_head, k_head, v_head, block_mask=mask)
> > ```
> > return hidden_states
> > ```
> 
> In Wanx, I only apply sta on self-attention in which qkv are both video, so it's more like stepvideo rather than hunyuan. Have you ever tried other numbers for tile_thw, kernel_thw? I find [#220](https://github.com/hao-ai-lab/FastVideo/pull/220) add some heuristic method to choose config.

kernel size在这里
https://github.com/hao-ai-lab/FastVideo/blob/main/assets/mask_strategy_hunyuan.json

tile size是6,8,10

**Comment by botbw** [comment]
> > [@shadow150519](https://github.com/shadow150519) hunyuan的是改成下面这样，stepvideo可以简单替换；但是hunyuan有has_text，flex_attention没办法直接简单完成；因为text这个要单独处理逻辑；
> > def flex_test(Q, K, V, window_size, has_text): hidden_states = torch.empty_like(Q) for head_index, (t_kernel, h_kernel, w_kernel) in enumerate(window_size): q_head, k_head, v_head, o_head = Q[:, head_index:head_index + 1], K[:, head_index: head_index + 1], V[:, head_index: head_index + 1], hidden_states[:, head_index: head_index + 1] mask = get_sliding_tile_attention_mask((**t_kernel, h_kernel, w_kernel), (6, 8, 10), (30, 48, 80)**, 39, 'cuda', 0) hidden_states[:,head_index:head_index + 1] = flex_attention(q_head, k_head, v_head, block_mask=mask)
> > ```
> > return hidden_states
> > ```
> 
> In Wanx, I only apply sta on self-attention in which qkv are both video, so it's more like stepvideo rather than hunyuan. Have you ever tried other numbers for tile_thw, kernel_thw? I find [#220](https://github.com/hao-ai-lab/FastVideo/pull/220) add some heuristic method to choose config.

@shadow150519 Have you tried increasing tile_size and window_size so that sta behaves almost like normal attention? One thing to notice is that: at the early stage of diffusion, sta always behaves like this (you can check `mask_strategy_xxx.json`).

**Comment by shadow150519** [comment]
> > > [@shadow150519](https://github.com/shadow150519) hunyuan的是改成下面这样，stepvideo可以简单替换；但是hunyuan有has_text，flex_attention没办法直接简单完成；因为text这个要单独处理逻辑；
> > > def flex_test(Q, K, V, window_size, has_text): hidden_states = torch.empty_like(Q) for head_index, (t_kernel, h_kernel, w_kernel) in enumerate(window_size): q_head, k_head, v_head, o_head = Q[:, head_index:head_index + 1], K[:, head_index: head_index + 1], V[:, head_index: head_index + 1], hidden_states[:, head_index: head_index + 1] mask = get_sliding_tile_attention_mask((**t_kernel, h_kernel, w_kernel), (6, 8, 10), (30, 48, 80)**, 39, 'cuda', 0) hidden_states[:,head_index:head_index + 1] = flex_attention(q_head, k_head, v_head, block_mask=mask)
> > > ```
> > > return hidden_states
> > > ```
> > 
> > 
> > In Wanx, I only apply sta on self-attention in which qkv are both video, so it's more like stepvideo rather than hunyuan. Have you ever tried other numbers for tile_thw, kernel_thw? I find [#220](https://github.com/hao-ai-lab/FastVideo/pull/220) add some heuristic method to choose config.
> 
> [@shadow150519](https://github.com/shadow150519) Have you tried increasing tile_size and window_size so that sta behaves almost like normal attention? One thing to notice is that: at the early stage of diffusion, sta always behaves like this (you can check `mask_strategy_xxx.json`).

Yes, I've check mask_strategy_hunyuan.json and find in early step，STA actually behave like full attention. I tried to remove attention_mask in flex_attention to do a full 3D attention and it can generate normal video. 
Because the latent vector size in my scenario is only (21, 30, 52), selecting tile_thw(2, 2, 4) and kernel_thw(4, 4, 8) effectively creates an attention window of (10, 10, 36). If these values are set too large, it seems like it wouldn't reduce computation significantly. However, with the current settings, it generates meaningless videos. Additionally, I noticed that in your paper, you need to configure different settings for each head of each layer individually. Perhaps the issue is that I've simply set the same values for every head of every layer. I'm going to analyze the distribution of the attention scores next

**Comment by botbw** [comment]
> > > > [@shadow150519](https://github.com/shadow150519) hunyuan的是改成下面这样，stepvideo可以简单替换；但是hunyuan有has_text，flex_attention没办法直接简单完成；因为text这个要单独处理逻辑；
> > > > def flex_test(Q, K, V, window_size, has_text): hidden_states = torch.empty_like(Q) for head_index, (t_kernel, h_kernel, w_kernel) in enumerate(window_size): q_head, k_head, v_head, o_head = Q[:, head_index:head_index + 1], K[:, head_index: head_index + 1], V[:, head_index: head_index + 1], hidden_states[:, head_index: head_index + 1] mask = get_sliding_tile_attention_mask((**t_kernel, h_kernel, w_kernel), (6, 8, 10), (30, 48, 80)**, 39, 'cuda', 0) hidden_states[:,head_index:head_index + 1] = flex_attention(q_head, k_head, v_head, block_mask=mask)
> > > > ```
> > > > return hidden_states
> > > > ```
> > > 
> > > 
> > > In Wanx, I only apply sta on self-attention in which qkv are both video, so it's more like stepvideo rather than hunyuan. Have you ever tried other numbers for tile_thw, kernel_thw? I find [#220](https://github.com/hao-ai-lab/FastVideo/pull/220) add some heuristic method to choose config.
> > 
> > 
> > [@shadow150519](https://github.com/shadow150519) Have you tried increasing tile_size and window_size so that sta behaves almost like normal attention? One thing to notice is that: at the early stage of diffusion, sta always behaves like this (you can check `mask_strategy_xxx.json`).
> 
> Yes, I've check mask_strategy_hunyuan.json and find in early step，STA actually behave like full attention. I tried to remove attention_mask in flex_attention to do a full 3D attention and it can generate normal video. Because the latent vector size in my scenario is only (21, 30, 52), selecting tile_thw(2, 2, 4) and kernel_thw(4, 4, 8) effectively creates an attention window of (10, 10, 36). If these values are set too large, it seems like it wouldn't reduce computation significantly. However, with the current settings, it generates meaningless videos. Additionally, I noticed that in your paper, you need to configure different settings for each head of each layer individually. Perhaps the issue is that I've simply set the same values for every head of every layer. I'm going to analyze the distribution of the attention scores next

@shadow150519 Sorry, I'm not the maintainer or author, we are doing similar things (and observed meaningless artifacts after using sta).

Non-supported models cannot function normally by simply replacing fla with sta. It seems like (I might be wrong) the hyper-param searching is still necessary as the config from current supported models cannot be applied directly.

**Comment by wangshankun** [comment]
@botbw @shadow150519  要不要加个微信群，一起讨论？ 我的号: TheWeChatOfShwan

我刚实现img和text的sp并行，计划将img2img使用STA类似大稀疏的，其他注意力img2text，text2img，text2text用Full Attention

**Comment by TumCCC** [comment]
> > > > [@shadow150519](https://github.com/shadow150519) hunyuan的是改成下面这样，stepvideo可以简单替换；但是hunyuan有has_text，flex_attention没办法直接简单完成；因为text这个要单独处理逻辑；
> > > > def flex_test(Q, K, V, window_size, has_text): hidden_states = torch.empty_like(Q) for head_index, (t_kernel, h_kernel, w_kernel) in enumerate(window_size): q_head, k_head, v_head, o_head = Q[:, head_index:head_index + 1], K[:, head_index: head_index + 1], V[:, head_index: head_index + 1], hidden_states[:, head_index: head_index + 1] mask = get_sliding_tile_attention_mask((**t_kernel, h_kernel, w_kernel), (6, 8, 10), (30, 48, 80)**, 39, 'cuda', 0) hidden_states[:,head_index:head_index + 1] = flex_attention(q_head, k_head, v_head, block_mask=mask)
> > > > ```
> > > > return hidden_states
> > > > ```
> > > 
> > > 
> > > In Wanx, I only apply sta on self-attention in which qkv are both video, so it's more like stepvideo rather than hunyuan. Have you ever tried other numbers for tile_thw, kernel_thw? I find [#220](https://github.com/hao-ai-lab/FastVideo/pull/220) add some heuristic method to choose config.
> > 
> > 
> > [@shadow150519](https://github.com/shadow150519) Have you tried increasing tile_size and window_size so that sta behaves almost like normal attention? One thing to notice is that: at the early stage of diffusion, sta always behaves like this (you can check `mask_strategy_xxx.json`).
> 
> Yes, I've check mask_strategy_hunyuan.json and find in early step，STA actually behave like full attention. I tried to remove attention_mask in flex_attention to do a full 3D attention and it can generate normal video. Because the latent vector size in my scenario is only (21, 30, 52), selecting tile_thw(2, 2, 4) and kernel_thw(4, 4, 8) effectively creates an attention window of (10, 10, 36). If these values are set too large, it seems like it wouldn't reduce computation significantly. However, with the current settings, it generates meaningless videos. Additionally, I noticed that in your paper, you need to configure different settings for each head of each layer individually. Perhaps the issue is that I've simply set the same values for every head of every layer. I'm going to analyze the distribution of the attention scores next

Hello, I am also trying to explore the performance of STA on WAN2.1, but I also get a bunch of noise with using original STA.cu implementation not the ref. I would like to ask how about your analysis of attention distribution of Wanselfattention?
